### PR TITLE
kafka replay speed: handle metadata refreshes and OffsetOutOfRange errors

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
#### What this PR does

1. Deploys 3 kafka brokers with 2x replication so that you can simulate electing a new kafka broker for a partition locally
2. Handle leader brokers in concurrent fetchers
   1. Selecting the leader broker to send a request to. Currently we don't support the preferred replicas that the kafka API provides.
   2. Handle errors around leader election and broker selection - this is basically triggering a metadata refresh whenever we see some specific errors. franz-go already refreshes metadata when it notices changes in brokers, but it's hard to tell whether this will work in all cases for us. Refreshing metadata has few (no) downsides.

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/storage/ingest/pusher.go [hunk 1]
```diff
@@ -339,6 +339,8 @@ func (p *shardingPusher) PushToStorage(ctx context.Context, request *mimirpb.Wri
 		s.Context = ctx // retain the last context in case we have to flush it when closing shardingPusher
 
 		if len(s.Timeseries) < p.batchSize {
+			// We modified a value so far. Write it back.
+			p.unfilledShards[shard] = s
```
Comment (dimitarvdimitrov) [inline pkg/storage/ingest/pusher.go:343]: this is unrelated. I found the bug 2 weeks ago for some reason, so now I'm commiting it.

Comment (gotjosh) [reply pkg/storage/ingest/pusher.go:343]: I don' think we need this since the introduction of https://github.com/grafana/mimir/pull/9133 - I fixed it there.

File: pkg/storage/ingest/reader.go [hunk 2]
```diff
@@ -807,6 +807,10 @@ type fetchResult struct {
 	fetchedBytes int
 }
 
+func emptyFetchResult(err error) fetchResult {
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:810]: nit: I feel like this prefix suits this better.
```suggestion
func newEmptyFetchResult(err error) fetchResult {
```

File: pkg/storage/ingest/reader.go [hunk 3]
```diff
@@ -853,6 +857,8 @@ func newConcurrentFetchers(ctx context.Context, client *kgo.Client, logger log.L
 	topics, err := kadm.NewClient(client).ListTopics(ctx, topic)
 	if err != nil {
 		return nil, fmt.Errorf("failed to find topic ID: %w", err)
+	} else if !topics.Has(topic) {
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:860]: nit: Why the nested else?

We could simply add another `if`

File: pkg/storage/ingest/reader.go [hunk 4]
```diff
@@ -894,45 +900,57 @@ func (r *concurrentFetchers) pollFetches(ctx context.Context) (result kgo.Fetche
 	}
 }
 
+var errUnknownPartitionLeader = fmt.Errorf("unknown partition leader")
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:903]: nit: should we place this sat the top next to the other errors?

File: pkg/storage/ingest/reader.go [hunk 5]
```diff
@@ -894,45 +900,57 @@ func (r *concurrentFetchers) pollFetches(ctx context.Context) (result kgo.Fetche
...
+
+// fetchMinBytesWaitTime is the time the Kafka broker can wait for MinBytes to be filled.
+// This is usually used when there aren't enough records available to fulfil MinBytes, so the broker waits for more records to be produced.
+var fetchMinBytesWaitTime = 10 * time.Second
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:907]: nit: I tend to discourage the use of global variables for methods that have a pointer receiver. You could simplify this and add this as an attribute to `concurrentFetchers` then rename this  to `defaultFetchMinBytesWaitTime` which is set on the constructor.

Also, non-nit can you please specify the unit in the name, it was non-obvious that even though this in seconds, the actual client requires millis.

Comment (gotjosh) [reply pkg/storage/ingest/reader.go:907]: Very similar to what you have done with `noReturnedRecords` which btw, I'd suggest you consider moving that to the top too.

File: pkg/storage/ingest/reader.go [hunk 6]
```diff
@@ -894,45 +900,57 @@ func (r *concurrentFetchers) pollFetches(ctx context.Context) (result kgo.Fetche
...
+
+// fetchSingle sends a fetch request to the leader Kafka broker for a partition for the fetchWant and parses the responses.
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:909]:
```suggestion
// fetchSingle attempts to find out the leader leader Kafka broker for a partition and then sends a fetch request to the leader of the fetchWant request and parses the responses
```

File: pkg/storage/ingest/reader.go [hunk 7]
```diff
@@ -1122,20 +1144,50 @@ func handleKafkaFetchErr(fr fetchResult, fw fetchWant, shortBackoff, longBackoff
...
+	// Typically franz-go will update its own metadata when it detects a change in brokers. But it's hard to verify this.
+	// So we force a metadata refresh here to be sure.
+	// It's ok to call this from multiple fetchers concurrently. franz-go will only be sending one metadata request at a time (whether automatic, periodic, or forced).
+	//
+	// Metadata refresh is asynchronous. So even after forcing the refresh we might have outdated metadata.
+	// Hopefully the backoff that will follow is enough to get the latest metadata.
+	// If not, the fetcher will end up here again on the next attempt.
+	client.ForceMetadataRefresh()
+}
+
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:1190]: nit: I don't think we need this level of interaction - if you care more about the comment than anything else, you can just inline the metadata refresh in a closure within `handleKafkaFetchErr`.

e.g.
```golang
func handleKafkaFetchErr(fr fetchResult, fw fetchWant, shortBackoff, longBackoff waiter, client *kgo.Client, logger log.Logger) fetchWant {
	refreshMetadata := client.ForceMetadataRefresh
```

and then add your comment on top, I do understand that this makes it simpler to test but for now, It feels like we should have a `ReaderClient` or something interface around here, which is kind of a different problem.

**Review by gotjosh — APPROVED** [review]
LGTM, almost all of my comments are nits.

File: pkg/storage/ingest/pusher.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/pusher.go:343]:
```suggestion
```

File: pkg/storage/ingest/reader.go (hunk 4 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go:903]: good spot, moved

File: pkg/storage/ingest/reader.go (hunk 5 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go:907]: > Also, non-nit can you please specify the unit in the name, it was non-obvious that even though this in seconds, the actual client requires millis.

it's a `time.Duration`, isn't the type enough? It's effectively in nanoseconds, but it's a proper type

Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go]: the reason i didn't want to do it in the config is because it requires a lot of prod code changes just to ease tests. Take a look at this - a404ec94168f8ee11858037ba42e6b8fb838e3fe - if you like it, we can keep it

File: pkg/storage/ingest/reader.go (hunk 7 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go:1188]: i do care about the comment and can inline this function, but what's wrong with the interface? we have the two `waiter`s too? for now we need only this if we later have a need for `ReaderClient` we can replace `metadataRefresher`, but we don't now (do we?)

File: pkg/storage/ingest/reader.go (hunk 5 above)
Comment (gotjosh) [reply pkg/storage/ingest/reader.go]: I was thinking much simpler than the above:

```git
diff --git a/pkg/storage/ingest/reader.go b/pkg/storage/ingest/reader.go
index 854dd0aaf..3bfbc3261 100644
--- a/pkg/storage/ingest/reader.go
+++ b/pkg/storage/ingest/reader.go
@@ -45,6 +45,8 @@ const (
 
 	// kafkaOffsetEnd is a special offset value that means the end of the partition.
 	kafkaOffsetEnd = int64(-1)
+
+	defaultMinBytesWaitTimeSeconds = 10 * time.Second
 )
 
 var (
@@ -840,7 +842,7 @@ func newConcurrentFetchers(ctx context.Context, client *kgo.Client, logger log.L
 		partitionID:        partition,
 		metrics:            metrics,
 		recordsPerFetch:    recordsPerFetch,
-		minBytesWaitTime:   minBytesWaitTime,
+		minBytesWaitTime:   defaultMinBytesWaitTimeSeconds,
 		lastReturnedRecord: noReturnedRecords,
 		tracer:             kotel.NewTracer(kotel.TracerPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}))),
 		orderedFetches:     make(chan kgo.FetchPartition),

```

File: pkg/storage/ingest/reader.go (hunk 7 above)
Comment (gotjosh) [reply pkg/storage/ingest/reader.go]: I was more concerned with the indirection than interface to be honest. I'm generally not a fan of single method interfaces but can appreciate it's very subjective, this LGTM.

**Review by gotjosh — APPROVED** [review]
LGTM, I don't need to look at this again.

File: pkg/storage/ingest/reader.go (hunk 5 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go]: oh, that works too. I prefer that. Done in a6b2e87c59aff7881a2a24ced5e2ee98407739c7

File: pkg/storage/ingest/reader.go [hunk 8]
```diff
@@ -829,8 +829,12 @@ type concurrentFetchers struct {
 	lastReturnedRecord int64
 }
 
+// defaultMinBytesWaitTime is the time the Kafka broker can wait for MinBytes to be filled.
+// This is usually used when there aren't enough records available to fulfil MinBytes, so the broker waits for more records to be produced.
+var defaultMinBytesWaitTime = 10 * time.Second
```
Comment (gotjosh) [inline pkg/storage/ingest/reader.go:834]: I think we need to move this next to the top of the fiile.

File: pkg/storage/ingest/reader_test.go [hunk 9]
```diff
@@ -30,6 +30,11 @@ import (
 	"github.com/grafana/mimir/pkg/util/testkafka"
 )
 
+func init() {
+	// Reduce the time the fake kafka would wait for new records. Sometimes this blocks startup.
+	defaultMinBytesWaitTime = time.Second
```
[the remark this answers is not in the document]
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader_test.go:35]: i think this should do it https://github.com/grafana/mimir/pull/9202

File: pkg/storage/ingest/reader.go (hunk 8 above)
Comment (dimitarvdimitrov) [reply pkg/storage/ingest/reader.go:834]: in #9202 it's now a constant and at the top
